external/qjs_crypto_module.c

```diff
@@ -0,0 +1,678 @@
+/*
+ * Copyright (C) NGINX, Inc.
+ *
+ * This module implements the crypto API for QuickJS.
+ *
+ * It is a port of njs_crypto_module.c file.
+ */
+
+#include <qjs.h>
+#include "njs_hash.h"
+
+typedef void (*njs_hash_init)(njs_hash_t *ctx);
+typedef void (*njs_hash_update)(njs_hash_t *ctx, const void *data, size_t size);
+typedef void (*njs_hash_final)(u_char result[32], njs_hash_t *ctx);
+
+typedef JSValue (*qjs_digest_encode)(JSContext *cx, const njs_str_t *src);
+
+
+typedef struct {
+    njs_str_t      name;
+
+    size_t         size;
+    njs_hash_init  init;
+    njs_hash_update update;
+    njs_hash_final final;
+} njs_hash_alg_t;
+
+typedef struct {
+    JSContext      *cx;
```

Comment:
We do not need `JSContext *` for `njs_digest_t ` and `njs_hmac_t `.
`js_free_rt()` can be used at finalizers.